File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/repository/query/AbstractRequeryQuery.java
package com.coupang.springframework.data.requery.repository.query;

import com.coupang.springframework.data.requery.core.RequeryOperations;
import com.coupang.springframework.data.requery.repository.query.RequeryQueryExecution.*;
import com.coupang.springframework.data.requery.utils.RequeryMetamodel;
import com.coupang.springframework.data.requery.utils.RequeryUtils;
import io.requery.query.Scalar;
import io.requery.query.Tuple;
import io.requery.query.element.QueryElement;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.data.repository.query.ParametersParameterAccessor;
import org.springframework.data.repository.query.RepositoryQuery;
import org.springframework.data.repository.query.ReturnedType;
import org.springframework.util.Assert;

import java.util.Optional;


/**
 * Abstract base class to implement {@link RepositoryQuery}s.
 *
 * @author debop
 * @since 18. 6. 7
 */
@Getter
@Slf4j
public abstract class AbstractRequeryQuery implements RepositoryQuery {

    protected final RequeryQueryMethod queryMethod;
    protected final RequeryOperations operations;
    protected final RequeryMetamodel metamodel;
    protected final Class<?> domainClass;

    public AbstractRequeryQuery(@NotNull final RequeryQueryMethod method,
                                @NotNull final RequeryOperations operations) {
        Assert.notNull(method, "queryMethod must not be null");
        Assert.notNull(operations, "operations must not be null");

        this.queryMethod = method;
        this.operations = operations;
        this.metamodel = new RequeryMetamodel(RequeryUtils.getEntityModel(operations.getDataStore()));
        this.domainClass = method.getEntityInformation().getJavaType();
    }

    public Object execute(Object[] parameters) {
        return doExecute(getExecution(), parameters);
    }

    @Nullable
    private Object doExecute(@NotNull RequeryQueryExecution execution, Object[] values) {

        Object result = execution.execute(this, values);
        ParametersParameterAccessor accessor = new ParametersParameterAccessor(queryMethod.getParameters(), values);

        log.debug("doExecute ... result={}", result);
        return result;
    }

    protected RequeryQueryExecution getExecution() {
        // NOTE: 검사하는 순서가 중요합니다.
        //
        if (queryMethod.isStreamQuery()) {
            log.debug("Create StreamExecution. queryMethod={}", queryMethod);
            return new StreamExecution(queryMethod.getParameters());
        } else if (queryMethod.isCollectionQuery()) {
            log.debug("Create CollectionExecution. queryMethod={}", queryMethod);
            return new CollectionExecution();
        } else if (queryMethod.isSliceQuery()) {
            log.debug("Create SliceExecution. queryMethod={}", queryMethod);
            return new SlicedExecution(queryMethod.getParameters());
        } else if (queryMethod.isPageQuery()) {
            log.debug("Create PagedExecution. queryMethod={}", queryMethod);
            return new PagedExecution(queryMethod.getParameters());
        } else if (queryMethod.isModifyingQuery()) {
            return new ModifyingExecution(queryMethod, operations);
        } else {
            log.debug("Create SingleEntityExecution. queryMethod={}", queryMethod);
            return new SingleEntityExecution();
        }
    }

    protected QueryElement<?> createQueryElement(Object[] values) {
        log.debug("Create QueryElement with domainClass={}, values={}", domainClass.getName(), values);
        return doCreateQuery(values);
    }

    protected QueryElement<? extends Scalar<Integer>> createCountQueryElement(Object[] values) {
        return doCreateCountQuery(values);
    }

    protected Optional<Class<?>> getTypeToRead(ReturnedType returnedType) {
        return returnedType.isProjecting() && !getMetamodel().isRequeryManaged(returnedType.getReturnedType())
               ? Optional.of(Tuple.class)
               : Optional.empty();
    }

    protected abstract QueryElement<?> doCreateQuery(Object[] values);

    protected abstract QueryElement<? extends Scalar<Integer>> doCreateCountQuery(Object[] values);


}

File: spring-data-requery/src/main/kotlin/org/springframework/data/requery/repository/query/AbstractRequeryQuery.kt
package org.springframework.data.requery.repository.query

import io.requery.query.Result
import io.requery.query.Tuple
import io.requery.query.element.QueryElement
import mu.KotlinLogging
import org.springframework.data.repository.query.QueryMethod
import org.springframework.data.repository.query.RepositoryQuery
import org.springframework.data.repository.query.ReturnedType
import org.springframework.data.requery.core.RequeryOperations
import org.springframework.data.requery.utils.RequeryMetamodel

/**
 * Abstract base class to implement [RepositoryQuery]s.
 *
 * @author debop@coupang.com
 */
abstract class AbstractRequeryQuery(protected val queryMethod: RequeryQueryMethod,
                                    protected val operations: RequeryOperations): RepositoryQuery {

    companion object {
        private val log = KotlinLogging.logger { }
    }

    protected val metamodel = RequeryMetamodel(operations.entityModel)
    protected val domainClass = queryMethod.entityInformation.javaType as Class<out Any>

    override fun getQueryMethod(): QueryMethod = queryMethod

    override fun execute(parameters: Array<Any?>): Any? = doExecute(getExecution(), parameters)

    private fun doExecute(execution: RequeryQueryExecution, values: Array<Any?>): Any? {

        val result = execution.execute(this, values)
        // 필요없는데 ???
        //        val accessor = ParamtersParameterAccessor(queryMethod.parameters, values)

        log.debug { "doExecute ... result=[$result]" }
        return result
    }

    protected open fun getExecution(): RequeryQueryExecution {

        return when {
            queryMethod.isStreamQuery -> StreamExecution(queryMethod.parameters)
            queryMethod.isCollectionQuery -> CollectionExecution()
            queryMethod.isSliceQuery -> SlicedExecution(queryMethod.parameters)
            queryMethod.isPageQuery -> PagedExecution(queryMethod.parameters)
            else -> SingleEntityExecution()
        }
    }

    protected abstract fun doCreateQuery(values: Array<Any?>): QueryElement<out Any>

    protected abstract fun doCreateCountQuery(values: Array<Any?>): QueryElement<out Result<Int>>

    internal fun createQueryElement(values: Array<Any?>): QueryElement<out Any> {
        log.debug { "Create QueryElement with domainClass=${domainClass.name}, values=$values" }
        return doCreateQuery(values)
    }

    protected fun createCountQueryElement(values: Array<Any?>): QueryElement<out Result<Int>> {
        return doCreateCountQuery(values)
    }

    protected fun getTypeToRead(returnedType: ReturnedType): Class<*>? {
        return when {
            returnedType.isProjecting && !metamodel.isRequeryManaged(returnedType.returnedType) -> Tuple::class.java
            else -> null
        }
    }
}